src/csharp/Grpc.Tools.MSBuildIntegration/Grpc.Tools.MSBuildIntegration.targets

```diff
@@ -0,0 +1,27 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <CoreCompileDependsOn>Proto;$(CoreCompileDependsOn)</CoreCompileDependsOn>
+  </PropertyGroup>
+  <ItemGroup>
+    <AvailableItemName Include="Proto"/>
+  </ItemGroup>
+  <Target Name="Proto" Inputs="@(Proto)" Outputs="@(Proto->'$(IntermediateOutputPath)%(FileName).cs')">
+    <!-- Run our platform independent wrapper that launches protoc with the right arguments -->
+    <Exec Condition="'$(OS)' == 'Windows_NT'" Command="$(MSBuildThisFileDirectory)..\tools\Grpc.Tools.CSharpGenerator.exe $(ProjectDir) $(IntermediateOutputPath) @(Proto->'%(filename)%(extension)', ' ')" LogStandardErrorAsError="true"/>
+    
+    <!-- Make sure that our protoc and protoc_csharp_plugin files are executable -->
+    <Exec Condition="'$(OS)' == 'Unix'" Command="chmod +x $(MSBuildThisFileDirectory)../tools/*/protoc"/>
+    <Exec Condition="'$(OS)' == 'Unix'" Command="chmod +x $(MSBuildThisFileDirectory)../tools/*/grpc_csharp_plugin"/>    
+
+    <!-- Run our platform independent wrapper that launches protoc with the right arguments -->
+    <Exec Condition="'$(OS)' == 'Unix'" Command="mono $(MSBuildThisFileDirectory)../tools/Grpc.Tools.CSharpGenerator.exe $(ProjectDir) $(IntermediateOutputPath) @(Proto->'%(filename)%(extension)', ' ')" LogStandardErrorAsError="true"/>
+    
+    <!-- Ensure that only files that exist (are touchable) are added in to the Compile target -->
+    <Touch Files="@(Proto->'$(IntermediateOutputPath)%(FileName).cs')" ForceTouch="true">

```

Comment:
Please also add the generated files into the <FileWrites> collection. These are the files that are removed when Clean and Rebuild targets are run. Otherwise, Rebuild will not rebuild the proto files.

Actually, just FYI, to keep the build hermetic, the build system saves the list of files generated during the last build, and removes the generated files based not on the current state of project, but on the state at the time the file were generated. This way, even if you remove x.cs from project, x.obj will be cleaned. It is one of .txt files in the intermediate directory.